### ISSUE TITLE
Allow option to use DataGeometry objects à la scikit-learn pipelines

### ISSUE THREAD

**Issue by paxtonfitzpatrick** [ContextLab/hypertools#227]
Currently, if you want to repeatedly transform text samples with `hypertools.tools.format_data()` using the same parameters, the function re-fits both the vectorizer and text model on each call.  This ends up being fairly inefficient, and for expensive/numerous operations, makes working directly with the underlying `sklearn` classes the better option.  

We could add an argument to return the fit models for reuse, but a really nice feature would be something like a [scikit-learn Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) object that you could create, fit, save, and reuse to perform various processing steps with a single call.  This would also be a very attractive feature for hypertools, since it could also additionally implement methods like `.plot()` and `.describe()`.

**Comment by jeremymanning** [comment]
Current status in the 1.0 update (PR #272): still not addressed, and arguably harder to build on than before -- `DataGeometry` (`hypertools/datageometry.py`) is now explicitly internal-only (its own docstring: 1.0 users never receive one; `plot()` returns a bare Figure, `load()` returns raw data), so it isn't a usable base for a public pipeline object. `format_data()`/`text2mat()` (`hypertools/tools/format_data.py`, `text2mat.py`) still re-fit the vectorizer/semantic model on every call; the only workaround is manually passing an already-fitted instance in place of a string.

**How this could be built on the 1.0 architecture:**
- Most of the plumbing already exists, just not assembled: `hypertools/core/model.py`'s `apply_model(..., return_model=True)` already fits once and returns the fitted model(s) for reuse; `hypertools/manip/common.py`'s `Manipulator` is already a proper `sklearn.base.BaseEstimator` (fit/transform/fit_transform) -- a good template to extend or subclass.
- Add a new public class (e.g. `hypertools/core/pipeline.py:Pipeline`) wrapping `normalize -> reduce -> align -> cluster`, mirroring `tools/analyze.py`'s existing functional chain but *keeping* each stage's fitted model instead of discarding it. `.fit(data)` / `.transform(new_data)` reuse those fitted stages; `.plot()` / `.describe()` delegate to the existing `hyp.plot()` / `hyp.describe()` on the transformed output.
- `reduce()`, `align()` (`tools/align.py`), and `cluster()` (`cluster/cluster.py`) don't currently expose `return_model` themselves (unlike `core.model.apply_model`, which shares their model registries) -- they'd need a thin passthrough, or the new class could call `apply_model` directly for the reduce/cluster stages.
- Align is the hard case: hyperalignment/SRM fitted transforms for genuinely new (held-out) data need explicit support and aren't free lunch. A v1 scoped to normalize+reduce+manip+cluster (deferring align-on-new-data) would be more tractable.

**Alternatives / prior art:**
- `sklearn.pipeline.Pipeline` is the direct model and is already used internally in `text2mat.py` for vectorizer->model chaining; hypertools' `Manipulator` base class already follows its fit/transform conventions.
- No third-party package wraps hypertools-style multi-dataset align/reduce/cluster with a fit-once/save/reuse/`.plot()`/`.describe()` API -- this is a hypertools-specific value-add, not an off-the-shelf adoption.

Note: parts of this ask (getting fitted models back out) are cheap given `apply_model`'s existing `return_model=True`; the harder, genuinely new part is a stateful public class plus align's held-out-data transform support.

Tagging as **medium effort**; leaving open for consideration in a future release.

**Comment by jeremymanning** [comment]
The Pipeline architecture is good; let's expose it to the user. we also need "return_model" flags for *all* hypertools modules. This should return either a single model (if just one step is applied) or a Pipeline object (if multiple steps). The Pipeline operator should then be able to be passed back to hyp.plot and hyp.analyze to apply the same pipeline to new data.

alignment can be applied *only* to new datasets of the same *total* size as the original-- e.g., we can learn a transform from a list of 20 items, each containing a 100 by 10 matrix, that projects the 20 10-dimensional trajectories into a common space. but that transform can only then be applied to new 20-item lists of n by 10 matrices.

**Comment by jeremymanning** [comment]
Addressed in PR #272 (HyperTools 1.0), per Jeremy's directive: expose the Pipeline architecture publicly, add `return_model=` to every hypertools module (returning a single model for one step, a `Pipeline` for multiple), and allow that Pipeline to be passed back to `hyp.plot`/`hyp.analyze` to apply the same fitted pipeline to new data. Alignment specifically may only be reapplied to new datasets of the same total size/shape as the originals — enforced, not just assumed.

New public `hyp.Pipeline` class (`hypertools/core/pipeline.py`, commit `b8ac8fe0`) with `build_pipeline()` assembling the canonical `manip → normalize → reduce → align → cluster` order. `return_model=False` was added to `reduce`, `cluster`, `align`, `normalize`, `manip`, and `analyze`; `plot()`/`analyze()` gained `pipeline=` to reuse a previously-fitted `Pipeline` on new data (commit `b49e70c2`, fix wave `db211e46`). `Aligner.transform(new_data)` now validates the new data's dataset count and per-dataset column count against fit time and raises `ValueError` naming the fit-time shape on mismatch, rather than silently replaying the original fit data (commit `084b5a61`).

```python
import hypertools as hyp

result, model = hyp.reduce(x, reduce='PCA', ndims=3, return_model=True)   # single model
result, pipeline = hyp.analyze(x, manip='Smooth', reduce='PCA', align='HyperAlign',
                                return_model=True)                        # Pipeline
fig = hyp.plot(new_data, pipeline=pipeline)                              # reapply, no refit
```

Verified: fitting `HyperAlign` on 3 rotated+noisy copies of a shared trajectory and transforming held-out data raises mean pairwise inter-dataset correlation from ~0.10 (unaligned) to ~0.995 (aligned); a shape-mismatched reuse call raises `ValueError` naming the fit-time dataset/column count. Poison-pill tests (monkeypatched `.fit` raises if called) confirm zero refit on reuse across `tests/test_pipeline.py` (23 tests) and `tests/test_align_migration.py` (28+ tests).

Full suite: ~1275 tests passing.
